src/ir/bits.h

```diff
@@ -128,35 +128,106 @@ struct DummyLocalInfoProvider {
 template<typename LocalInfoProvider = DummyLocalInfoProvider>
 Index getMaxBits(Expression* curr,
                  LocalInfoProvider* localInfoProvider = nullptr) {
-  if (auto* const_ = curr->dynCast<Const>()) {
+  if (auto* c = curr->dynCast<Const>()) {
     switch (curr->type.getSingle()) {
       case Type::i32:
-        return 32 - const_->value.countLeadingZeroes().geti32();
+        return 32 - c->value.countLeadingZeroes().geti32();
       case Type::i64:
-        return 64 - const_->value.countLeadingZeroes().geti64();
+        return 64 - c->value.countLeadingZeroes().geti64();
       default:
         WASM_UNREACHABLE("invalid type");
     }
   } else if (auto* binary = curr->dynCast<Binary>()) {
     switch (binary->op) {
       // 32-bit
-      case AddInt32:
-      case SubInt32:
-      case MulInt32:
-      case DivSInt32:
-      case DivUInt32:
-      case RemSInt32:
-      case RemUInt32:
       case RotLInt32:
       case RotRInt32:
+      case AddInt32:
+      case SubInt32:
+        return 32;
+      case MulInt32: {
+        auto maxBitsRight = getMaxBits(binary->right, localInfoProvider);
+        if (maxBitsRight == 32) {
+          return 32;
+        }
+        if (maxBitsRight == 0) {
+          return 0;
```

Comment:
are these checks for `maxBits == 0` useful? 0 means that the number is 0. In that case the optimizer should already emit a constant 0 in that place (if not, we should in one place check if `maxBits == 0` and replace it with a const there, instead of checking in multiple other places).

Reply:
yeah, it seems we could eliminate such checks